### ISSUE TITLE
[no-unsafe-readonly-mutable-assignment] false error with (mutable & readonly) type

### ISSUE THREAD

**Issue by Perfectoff** [danielnixon/eslint-plugin-total-functions#576]
This code throws an error when trying to assign (readonly & mutable) to readonly. I think there should be no error here.
```
type T_Mutable = { value : number[] | {a : number[]} };

type T_Readonly= { readonly value : readonly number[] | {readonly a : readonly number[]} };

declare const a : T_Mutable & T_Readonly;

const b : T_Readonly = a; // ESLint error: Using a readonly type to initialize a mutable type can lead to unexpected mutation in the readonly value.
```

**Comment by danielnixon** [comment]
Thanks for reporting this one @Perfectoff.

I'm going to close this as won't fix because it's functioning as intended. Let me explain.

Let's start with this slightly simplified form of your example:

```ts
type T_Mutable = { value: number[] };

type T_Readonly = { readonly value: readonly number[] };

const a: T_Mutable & T_Readonly = { value: [1, 2, 3] };
```

`a` is indeed mutable. If I wanted to, I could do this:

```ts
// this compiles, a is mutable
a.value.pop();

// this compiles too, so a is mutable for two reasons
a.value = [0];
```

Let's say we permitted assignment from a `T_Mutable & T_Readonly` to a `T_Readonly`:

```ts
// Let's say we permitted this
// eslint-disable-next-line total-functions/no-unsafe-mutable-readonly-assignment
const b: T_Readonly = a;
```

`b` is readonly, so the goal of this rule is to ensure that it can't be changed in surprising ways.

```ts
// At any point in my program, I should observe the same value here. If it changes, that is surprising because `b` said it was readonly.
console.log(b.value); // [ 1, 2, 3 ]
```

Now, let's mutate `a` and see what happens to `b`:

```
a.value.pop();

// That's surprising, b's value changed even though it is readonly
console.log(b.value); // [ 1, 2 ]
```

The goal of this rule is to prevent surprising changes like that, so it's working as intended.
